### DISCUSSION TITLE
Switch focus of user to a different room/topic

### DISCUSSION THREAD

**Discussion by square-foot** [balzack/databag#66] [Ideas]
I am so happy with this project. I had one idea where a user is made to focus his attention to a particular topic (which in my lingo, and other chat's lingo is a "room") Is there some API call or something I can do from outside so that a particular user goes to a particular topic/room?  Thanks a lot for the time and effort that is being put in for this project

**Comment by balzack** [comment]
Thanks for your interest in the project! Can you tell me a little more about the feature? Are you suggesting a link that brings a person directly to a 'room', and it doesn't contain a left bar for the other 'rooms'? Can you describe the use case a little more? Thank you.

**Reply by square-foot** [reply]
Thank you so much for your kind attention.

Let me explain my situation: Say there is a group of students who are being taught several subjects by a teacher/s -- In the Databag chat system, there are topics for each subject: "history", "geography", "science" etc.  One way to transfer the attention of the student to a particular topic is to send a message to that topic. For e.g. the teacher may send a message to the "history" topic and say: "Class, let's now work here" But that is a manual process and requires the effort of the said teacher. 

Ideally i would want the students to automatically be drawn to the "history" topic due to some other event happening in the system.  For e.g. it may be possible to call an API at the backend of the chat and make the student get focused to the "history" topic -- no manual sending of a message just to get students to be focused to another topic/room.

If this is not possible, is there any BOT or something that can send a message to the appropriate topic, so that that would achieve a similar objective. For e.g. The BOT can send this message "You are now working on 'history' " 

Regarding the left bar with the list of topics/rooms -- that can continue to be visible, if hiding the left-bar is too difficult.  But once the above auto-focusing system is implemented, and there is an option in the configuration of Databag to hide the list of topic/rooms, it would be great.  Because then topic switching happens due to other events happening in the overall system where Databag is embedded.

In short, the databag administrator pre-decides/configures the list of topic/rooms users can chat on. It is like putting blinkers so that chats are always focused on pre-decided topics. 

In such a case, the only topics that can be created apart from the above pre-decided topics would be for person-to-person (i.e. DM) Those person-to-person topic names can be created as needed, so that people can directly message one another (if so required) But if this exception is tough, then I am willing to let go of DMs. After all; in the above example of teachers and students, it is not really required that students talk privately to each other.

**Reply by balzack** [reply]
Hello, I've spent some time thinking about how to support the features you are asking about. I think the full feature set you are asking for wouldn't require any backend changes, but would be a significantly different client. My vision for the system is to support a variety of different clients, so that's not really bad news. I will spend some more time thinking about what effort that will take but might be awhile before I can get to it. I was planning on adapting the backend to OpenWRT as my next big development effort.

Implementing a bot type feature would likely be much less effort and possibly provide the necessary functionality in the interim.  I am not quite sure how it would solve the feature you are asking for as it wouldn't take any action on the student's clients, but simply cause a message to be sent in the appropriate room.

Thanks again for your interest in the project. I will continue to think about how to achieve your primary feature request. Ideally it would be a client that includes other related features important for people using this type of application.

**Reply by square-foot** [reply]
Thank you so much for going through this and your efforts you are taking. I will also (with your permission, and if I have time) go through your project's client side and see if I can be of some help.  

I have always been looking for this capability. I am so grateful that you are considering it -- because usually chat applications seem to go through a "formula" approach for their UI --- a list of rooms/channels/topics on one side, and people electing to go into one of those items in the list and  clicking on it to go and chat there.  

But it does not suit applications where the chatting happens to be just one of the functionality that the entire solution needs. In several of my projects, I am would like to do such "chat context switching" from outside the chat sub-system, instead of allowing the user to elect to manually switch the context -- the teacher/student example was one of those.

Regarding the "bot" sending a message - I am assuming that because the bot sends the message to a particular subject; the student is also automatically drawn into that topic i.e. the UI has changed to that subject. If the "bot" sends a message and if that remains as a "notification" then I agree, it may not be of much help --as the student is still stuck in some other topic and would have to manually access the notification list and elect to go that topic aka subject the bot was drawing attention to.

When I try to embed or use chat systems such as slack/mattermost/zulip, etc into my own solutions; the chat actually degrades the overall user experience because lot of times my users are forced to do such topic selections -- thus wasting time and it also results in attention loss.  Hence I've never been able to really embed those "formulaic" chat solutions. 

I am so grateful that you do understand the situation I am in. Thank you very much

**Reply by balzack** [reply]
Please feel free to take a look at the code. The project implements a node that makes of a decentralized network. It implements an API that essentially does efficient synchronization of data. The data is organized into three levels which I have named 'channel', 'topic', 'tag', but they really are just names. In the current chat app, the 'channel' makes up a room and the 'topic' make up the messages, but for your system you probably would want the 'channel' to be the school, 'topic' to be the class/subject, and 'tag' to be the messages. The payload of each data item is open, and to be defined by the participating applications. The 'school' record can be updated to reflect which 'subject' is active for example. It's probably a bit hard to follow, but perhaps will make more sense looking at the code.

I see the value for some applications to direct the users to different discussion conversation, but I can see challenges in the usability. As a user, I would be a bit frustrated if I were in the middle of a discussion and suddenly I am in a whole new discussion. Since you mention other projects, I imagine you have already thought through this. The functionality you describe seems to be only a component of a larger educational project. I imagine you are looking for a way to integrate the communication functionality into one of your projects vs using a standalone admin directed chat application.

Let me know if you have questions. I will keep thinking about this as I continue my OpenWRT efforts.

**Reply by square-foot** [reply]
Thank you once again for your reply.  I will surely take a look at the code -- I just hope to to get enough time. Currently, I have only three people working for me: I, me and myself :-) and often these three pull me into many other areas of attention :-) 

Regarding the UI difficulty: I understand your perspective: a person who is chatting under one "topic/subject" suddenly get pulled away... 

However, in all the projects I am working, that situation never comes. For example; in the teacher/student example; when the student is drawn into a different subject -- the student would be expecting that to happen anyway. Similarly, in other projects too the chat participants are agreeing to such "pulling" away. 

In fact, what I am talking of is kind of a "super set" of what normally happens in most chat applications.  Let me name the API (or the BOT activity in the background) that I am requesting as a "context switcher"  All I am asking is to make the "context switcher" independent of the UI event that calls it.  I am suggesting breaking of the umbilical chord which is tying the context switching to the user's clicks on the list of topics; and let that happen as per the needs of the specific application where the chat is a part of.

Let me explain.

For example;  in the earlier teacher+students example; the context switching happened because some activity being done by the teacher pulled the students to another topic. 

Here is the conventional example: If one were to implement a conventional chat UI, with the left side showing a list of possible topics, then the user himself clicking on the topic he wants to chat in, invokes  the "context switcher".  So in this case, the context switching is indeed being done by the user himself.

I can think of many other situations -- in fact a lot more situations than where current conventional chat UI works where my suggestion can be useful (Pardon me, I am biased). 

For example; a regular website has 3 different tabs open with 3 different pages of the website.  One tab for pricing, the other tab for product and third tab about the API docs of the product. Now let's think of the chat system deployed here. When the visitor is in "pricing" page, the chat system is naturally switched to "pricing" topic. When the visitor shifts to "product" page, the chat is now in "product" topic, and when the visitor shifts to "API docs page"  the chat is now in "API docs" -- one could use different background color or some other differentiation in the chat UI to indicate the respective topic that the visitor is now in.

There are many more such situations: Think of a graphical application like an adventure game -- where the visitor moves into different rooms of the game. The visitor's chat will automatically get switched to that topic which is part of that adventure game. Long time back (I am 62 years) there was a game called "Where in the World Is Carmen Sandiego?" (released in the eighties) where a detective (BTW, which is the game player himself) is solving a crime by going to various cities following a trail of clues. There was no chat system then, but now I can think of a chat window, as follows: Imagine a different local assistant of the main detective , is stationed at each city. 

And it would be that respective assistant who would be automatically available in the chat, without you having to "select" that topic when you happen to visit that city. 

... and many, many more such applications. As I said, my suggestion actually encompasses many more possibilities and also _INCLUDES_ the current conventional UI seen in slack/mattermost/ etc
